# Documentation des scalers

# 1. MinMaxScaler

## Introduction
Ce notebook présente la documentation complète du MinMaxScaler, 
une technique de normalisation des données utilisée dans les 
pipelines de Machine Learning. Vous y trouverez la description, 
la formule, le fonctionnement de chaque méthode, des exemples 
de code et une comparaison avec Scikit-learn.

## Description
Le  MinMaxScaler est utilisé pour normaliser les données en ramenant 
chaque dans un intervalle [0, 1] grâce à la formule 
(X - min) / (max - min). cela permet d'éviter qu'une feature avec 
de grandes valeurs domine les autres lors de l'entraînement du modèle.

### Formule
X' = (X - min) / (max - min)

Pour accomplir ce travail, 07 méthodes au total ont été mis en places:
fit(); transform(); fit_transform() inverse_transform(); convert_to array() et
convert_to_back 

### Méthode fit()
La méthode fit() calcul et mémorise les valeurs maximales et minimale de chaque
feature a partir des donnése d'entrainement.La méthode fit() prend en entrée un tableau de données X, calcule et mémorise la valeur minimale et maximale de chaque colonne(feature) et retourne l'objet scaler pour permettre le chainage des méthodes.Elle doit obligatoirement etre appelée avant transform().

### Méthode Transform()
La méthode transform() a pour role de normalisée les données dans l'intervalle [0,1]. Pour cela, elle prend en entrée un tableau de données X, vérifie d'abord que fit() a bien été appelé (sinon elle lève une erreur ValueError). Ensuite, elle applique la formule (X - min) / (max - min) en utilisant les valeurs mémorisées, et retourne les données transformées dans l'intervalle [0,1] dans le meme format que l'entrée(DataFrame, Serie ou array).


### Méthode fit_transform()
La méhode fit_transform() représente a lui seul la réunion des deux premières méthodes.
Son role est de calculée les min/max et de normalisée en une seule étape. Pour cela,elle prend en entrée un tableau de donnée X, appelle d'abord fit(X) qui calcul et mémorise les valeurs minimale et maximale de chaque colonne(feature), puis appelle mmédiatement transform(X) qui applique la formule X' = (X - min) / (max - min) sur ces données, et retourne les données normalisées dans l'intervalle [0,1] dans le meme format que l'entrée.

### Méthode inverse_transform()
La méthode inverse transform() a pour role de reconstruit les donnés originales a partir des donnés normalisés.Pour cela, elle prend en entrée u tableau de donnée déja normalisées dans l'intervalle [0,1], vérifie que fit() a bien été appelé (sinon elle lève une erreur ValueError), puis applique la formule inverse min + X_scaled * (max - min) pour restituer les données dans leur échelle originale, et retourne le résultat dans le meme format que l'entrée.

### Méthode _convert_to_array()
La méthode _convert_to_array() a pour role de convertit les données d'entrée en tableau numpy pour que les calculs puissent etre éffectués uniformément. Elle prend en entrée des données X et les convertit en tableau numpy pour permettre les calculs. Si X est un dataframe, elle mémorise les noms de colonnes et retourne les valeurs sous formes de tableau.Si X est une séries, elle mémorise le nom de la colonne et retourne les valeurs. Dans les autres cas, elle convertit directement X en tableau numpy. Elle retourne toujour un numpy array.

### Méthode _convert_back()
La méthode _convert_back() est une méthode interne qui a pour role de reconvertit depuis numpay array vers le format d'origine. Elle prend en entrée le tableau numpay transformé et les données originales comme référence de format. elle reconvertit le tableau numpay vers le format originale: Si les données originales étaitent un dataframe, elle recrée un dataframe avec les noms de colonnes mémorisés. Si c'était une séries, elle recréé une série avec le nom de colonne mémorisé;sinon elle retourne le tableau numpay tel quel. 







Voici un exemple concret de l'utilisation du MinMaxScaler et une comparaison faites avec scikit-learn:

In [1]:
import sys
sys.path.append(r'C:\Users\LENOVO\ifri_mini_ml_lib\ifri_mini_ml_lib\preprocessing\preparation\scaler')

from min_max_scaler import MinMaxScaler
from standard_scaler import StandardScaler

print("Import réussi !")


Import réussi !


In [2]:
import numpy as np
import pandas as pd
import time
from sklearn.preprocessing import MinMaxScaler as SklearnMinMaxScaler

# Chargement du dataset
df = pd.read_csv('base3.csv')

# Sélection de 3 colonnes numériques
data = df[['Age', 'EstimatedSalary', 'Purchased']]

print("Dataset original :")
print(data.head(10))

### Notre implémentation
debut = time.time()
notre_scaler = MinMaxScaler()
notre_resultat = notre_scaler.fit_transform(data)
notre_temps = time.time() - debut

### Scikit-learn
debut = time.time()
sklearn_scaler = SklearnMinMaxScaler()
sklearn_resultat = sklearn_scaler.fit_transform(data)
sklearn_temps = time.time() - debut

print("Notre MinMaxScaler :")
print(notre_resultat)
print(f"Temps : {notre_temps:.6f} secondes")

print("\nScikit-learn MinMaxScaler :")
print(sklearn_resultat)
print(f"Temps : {sklearn_temps:.6f} secondes")

print("\nRésultats identiques :", np.allclose(notre_resultat, sklearn_resultat))

Dataset original :
   Age  EstimatedSalary  Purchased
0   19            19000          0
1   35            20000          0
2   26            43000          0
3   27            57000          0
4   19            76000          0
5   27            58000          0
6   27            84000          0
7   32           150000          1
8   25            33000          0
9   35            65000          0
Notre MinMaxScaler :
          Age  EstimatedSalary  Purchased
0    0.023810         0.029630        0.0
1    0.404762         0.037037        0.0
2    0.190476         0.207407        0.0
3    0.214286         0.311111        0.0
4    0.023810         0.451852        0.0
..        ...              ...        ...
395  0.666667         0.192593        1.0
396  0.785714         0.059259        1.0
397  0.761905         0.037037        1.0
398  0.428571         0.133333        0.0
399  0.738095         0.155556        1.0

[400 rows x 3 columns]
Temps : 0.001113 secondes

Scikit-learn MinMaxS

Voici l'analyse. Ajoute une cellule Markdown :
## Analyse des résultats

### 1. Résultats de la transformation
- Les deux scalers produisent des **résultats identiques (True)**
- Les données sont bien normalisées dans l'intervalle **[0, 1]** :
  - Age : **0.023** (19 ans) → **0.785** (60 ans)
  - Salaire : **0.029** (15000) → **0.451** (76000)
  - Purchased : **0.0** (non acheté) | **1.0** (acheté)

### 2. Performances temporelles
- Notre MinMaxScaler : **0.001113 secondes**
- Scikit-learn : **0.008728 secondes**
- Notre implémentation est environ **8 fois plus rapide**

### 3. Comparaison : MinMaxScaler vs StandardScaler

**MinMaxScaler :**
- Ramène les valeurs entre **0 et 1**
- Formule : **(X - min) / (max - min)**
- Très sensible aux valeurs aberrantes
- Bornes fixes **[0, 1]**
- À utiliser quand **les bornes des données sont connues**

**StandardScaler :**
- Centre les valeurs autour de **0**
- Formule : **(X - moyenne) / écart-type**
- Moins sensible aux valeurs aberrantes
- Pas de bornes fixes
- À utiliser quand **les données suivent une distribution normale**

### 4. Conclusion
- Notre MinMaxScaler produit des résultats corrects
- Il est **8 fois plus rapide** que Scikit-learn sur ce dataset de 400 lignes
- C'est une alternative fiable pour la normalisation des données dans les pipelines ML
